pandas/io/formats/csvs.py

```diff
@@ -75,50 +76,93 @@ def __init__(
         self.should_close = ioargs.should_close
         self.mode = ioargs.mode
 
+        # GH21227 internal compression is not used for non-binary handles.
+        if (
+            self.compression
+            and hasattr(self.path_or_buf, "write")
+            and "b" not in self.mode
+        ):
+            warnings.warn(
+                "compression has no effect when passing a non-binary object as input.",
+                RuntimeWarning,
+                stacklevel=2,
+            )
+            self.compression = None
+
         self.sep = sep
         self.na_rep = na_rep
         self.float_format = float_format
         self.decimal = decimal
-
         self.header = header
         self.index = index
         self.index_label = index_label
-        if encoding is None:
-            encoding = "utf-8"
-        self.encoding = encoding
         self.errors = errors
-
-        if quoting is None:
-            quoting = csvlib.QUOTE_MINIMAL
-        self.quoting = quoting
-
-        if quoting == csvlib.QUOTE_NONE:
-            # prevents crash in _csv
-            quotechar = None
+        self.quoting = quoting or csvlib.QUOTE_MINIMAL
         self.quotechar = quotechar
-
         self.doublequote = doublequote
         self.escapechar = escapechar
-
         self.line_terminator = line_terminator or os.linesep
-
         self.date_format = date_format
+        self.cols = cols
+        self.chunksize = chunksize
+
+    @property
+    def index_label(self):
```

Comment:
Done.
6041666d0 TYP: add type annotations as reviewed

Reply:
ba353a5 - here I created IndexLabel type annotation in ``pandas._typing`` so that there is no code duplication in to_csv definition and CSVFormatter definition in regards to ``index_label`` type.